Pal 869 user service auditing (#215)
gchq/Palisade-services: user-service/README.md

<!---
Copyright 2020 Crown Copyright

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
--->

# <img src="../logos/logo.svg" width="180">

# User Service

The User Service is responsible for providing the other Palisade components with knowledge of the users of the system. 
Note that we are avoiding having a separate notion of "Palisade" users that are distinct from the normal user accounts of the system. There is no such thing as a "Palisade" user.  

> One of the purposes of this service is to allow Palisade to adopt whatever notion of *user* the host environment has. For example, this may from a central directory service such as LDAP, 

Reviewer: ```suggestion
One of the purposes of this service is to allow Palisade to adopt whatever notion of *user* the host environment has. For example, this may come from a central directory service such as LDAP, 
```

Reviewer 2: Done

> One of the purposes of this service is to allow Palisade to adopt the whatever notion of *user* the host environment has. For example, this may from a central directory service such as LDAP, 

Reviewer 3: ```suggestion
One of the purposes of this service is to allow Palisade to adopt whatever notion of *user* the host environment has. For example, this may from a central directory service such as LDAP, 
```

host operating system account or PKI based user authentication.
The User Service separates this concern from the rest of the system. Other components use this service's API to request user details. 
Some deployments may also allow Palisade to add users to the system, hence the presence of the `addUser()` method in the `UserService` interface.

## Message Model and Database Domain

| UserRequest     | UserResponse     | AuditErrorMessage | AuditableUserResponse |
|:----------------|:-----------------|:------------------|:----------------------|
| *token          | *token           | *token            | *token                | 
| userId          | userId           | userId            | UserResponse          |  
| resourceId      | resourceId       | resourceId        | AuditErrorMessage     |
| context         | context          | context           |                       |
|                 | User             | exception         |                       | 
|                 |                  | serverMetadata    |                       |
  
(fields marked with * are acquired from headers metadata)

> The service accepts a `UserRequest` and a token, then checks that the User is in the cache; the technology of which is chosen in the User Services application.yaml and Spring profiles.

Reviewer: ```suggestion
The service accepts a `UserRequest` and a token, then checks if the User is in the cache; the technology of which is chosen in the User Services application.yaml and Spring profiles.
```

Reviewer 2: Done

> If the User exists in the cache, then it, alongside any attributes are retrieved, and used to create a `UserReponse` object which is then packaged into a `AuditableUserResponse` object with no errors, 

Reviewer: ```suggestion
If the User exists in the cache, then it, alongside any attributes are retrieved, and used to create a `UserReponse` object which is then packaged into a `AuditableUserResponse` object with no errors, 
```
```suggestion
If the User exists in the cache, then it, alongside any additional attributes, are retrieved, and used to create a `UserReponse` object which is then packaged into a `AuditableUserResponse` object with no errors.
```

Reviewer 2: Done

> and then this is sent via kafka, to the Resource Service for further processing. 

Reviewer: ```suggestion
Then, this is sent via Kafka to the [Resource Service](../resource-service) for further processing. 
```

Reviewer 2: Done

> If the User does not exist in the cache, then a `NoSuchUserIdException` is thrown within the Service, which is caught, and packaged in the `AuditableUserReponse`, alongside the original `UserRequest` and an `AuditErrorMessage`, 

Reviewer: If it isn't found in the cache, it calls the real service, which should throw a `NoSuchUserIdException`. Only our 'trivial' implementation, the NullUserService, always throws this exception.

Reviewer 2: Done

> this is then sent to the Audit Service and logged appropriatly.

Reviewer: ```suggestion
this is then sent to the [Audit Service](../audit-service) and audited appropriately.
```

Reviewer 2: Done

## REST Interface

> The application exposes two REST endpoints used for debugging or mocking kafka entrypoints:

Reviewer: "exposes two" but have only described one (`api/user/multi` is missing)

Reviewer 2: Done

> * `POST api/user`

Reviewer 4: We are not posting a user, but a request. The endpoint should be `api/users/requests` to keep in line with RESTful semantics.

Author: This would require a change to the other services also, as they follow the same convention of /api/*service e.g /api/mask
if its worth doing, we can create a ticket on the backlog

Reviewer 4: I think it is worth doing, but is a low priority. An issue in the backlog would be best I think.

Reviewer: I like the usage we currently have as it is meant for pushing items onto kafka queues. In this way, the REST endpoint reflects the name of the topic.

  - accepts an `x-request-token` `String` header, any number of extra headers, and an `UserRequest` body
  - returns a `202 ACCEPTED` after writing the headers and `AuditableUserResponse` to kafka

## Example JSON Request
```
curl -X POST user-service/api/user -H "x-request-token: test-request-token" -H "content-type: application/json" --data \
'{
   "userId": "test-user-id",
   "resourceId": "/test/resourceId",
   "context": {
     "class": "uk.gov.gchq.palisade.Context",
     "contents": {
       "purpose": "purpose"
     }
   }
 }'
```

## Example JSON Response
```
'{
  "userId": "test-user-id",
  "resourceId": "/test/resourceId",
  "context": {
    "class": "uk.gov.gchq.palisade.Context",
    "contents": {
      "purpose": "purpose"
    }
  },
  "user": {
    "userId": {
      "id": "test-user-id"
    },
    "roles": [
      "roles"
    ],
    "auths": [
      "auths"
    ],

>     "class": "uk.gov.gchq.palisade.User"

Reviewer 4: Should not be returning class. This is exposing the service internal model, which is subject to change.

Author: the rest post is an internal api that is only meant for debugging. Inside a cluster these endpoints wont be visible

Reviewer 4: Ok. We can let this go then.

Reviewer: If anything, you need this class as it may be a `uk.gov.gchq.palisade.example.ExampleUser` or some other richer object with extra fields

If the User does not exist in the cache then the request gets directed to the local implementation of the User service. If the User service implementation returns a User then a 
`UserResponse` object gets created and added to an `AuditableUserResponse` object which is sent to the [Resource service](../resource-service).
Should any errors get thrown when trying to access a User, for example a `NoSuchUserIdException`, then this information creates an `AuditErrorMessage` 
that is then sent to the [Audit service](../audit-service) to be audited appropriately.

## REST Interface

The application exposes two REST endpoints used for debugging or mocking kafka entrypoints:
* `POST api/user`
  - accepts an `x-request-token` `String` header, any number of extra headers, and a single `UserRequest` in the body
  - returns a `202 ACCEPTED` after writing the headers and `AuditableUserResponse` to kafka
* `POST api/user/multi`
  - accepts an `x-request-token` `String` header, any number of extra headers, and a list of `UserRequest`s within the body

>   - returns a `202 ACCEPTED` after writing the headers and `AuditableUserResponse` to kafka

Reviewer: It writes a UserRequest to kafka and returns, no?

Reviewer 2: You're quite right. Updated so that a UserRequest is sent to Kafka :D
